Rework offline image serving

slaperche-scality opened the pull request:
**Component**: buildchain

**Context**:

To support fully-offline installations, we currently provide a Docker *registry* in a deployment, and use `docker pull`/`docker save` to generate image tarballs, which are then pushed into the registry at deployment time.

This has several drawbacks:
- `docker pull/save` to (basically) fetch files from the internet is not necessarily efficient
- Due to `docker save`, layers that are shared across multiple images still end up duplicated multiple times in the various tarballs (which e.g. in case of the CentOS base image means duplicating roughly 200MB several times), inflating the size of the distribution ISO
- The registry image needs to be maintained
- Due to the images being copied from the ISO into the registry, this adds extra space requirements to the bootstrap node
- Currently, the deployed registry is writable by anybody :fearful: 

Ideally, we should be able to deploy a (fake) registry using the `nginx` instance also used to distribute RPM packages/Yum repositories that's part of the bootstrap infrastructure, serving files directly from the mounted ISO.

**Summary**:

In the buildchain:
- use `skopeo` instead of `docker` to save the images
- use hard links to deduplicate the images' layers
- with the  `static-container-registry.py` script, generate an `nginx` config to serve the images

During deployment:
- mount the new configuration files and the images directory into the `nginx` container
- inject the `nginx` image using `ctr cri load` (to avoid chicken or egg problem).
- add the new configuration files as annotation to the pod manifest (to ensure restart on changes)
- get rid of the existing registry

**Acceptance criteria**:

Bootstrap should be working (so you can run a `doit.sh vagrant_up` and check if everything is green).

---

Fixes: #1049

slaperche-scality committed:
build/image: use skopeo instead of docker

In this PR we replace (most of) the use of docker by skopeo.
This concerns the creation of both local and remote images.

One remaining use of docker is to generate the `pause` image, because
`ctr cri load` can only load images saved by docker.
We will tackle this use case in a dedicated commit so that we can use
skopeo here as well.

Note that the images are no longer saved as tar archive, but as flat
directory (which will allow us to deduplicate the layers in a later
commit), thus the `Image` class can no longer inherits from the
`FileTarget` classe (because it doesn't produce a single file as
output).

This is not an issue as all image but one (`pause` again) don't rely on
this inheritance. For the last one, we introduce the `RemoteTarImage`
class.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
lint: fix pylint configuration

Don't take import statements into account when searching for
similarities.

Otherwise we end up with warnings like that:

    ************* Module buildchain.targets.remote_image
    R:  1, 0: Similar lines in 2 files
    ==buildchain.image:36
    ==buildchain.packaging:31
    from buildchain import config
    from buildchain import constants
    from buildchain import coreutils
    from buildchain import targets
    from buildchain import types
    from buildchain import utils

     (duplicate-code)

which we don't want.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/image: add layer de-duplication

Several images have common layers and we can reduce the space used by
those layers by keeping a single copy of each of them (thanks to hard
links).

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1049-rework-offline-image-serving` into
`development/2.0`
* [w/2.1/feature/1049-rework-offline-image-serving](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1049-rework-offline-image-serving) will be merged into `development/2.1` 
* [w/2.2/feature/1049-rework-offline-image-serving](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1049-rework-offline-image-serving) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

NicolasT commented on buildchain/buildchain/image.py:
```diff
@@ -244,7 +260,6 @@ def task__image_build() -> Iterator[types.TaskDict]:
         version='2.7.1',
         digest='sha256:870474507964d8e7d8c3b53bcfa738e3356d2747a42adad26d0d81ef4479eb1b',
         destination=ISO_IMAGE_ROOT,
-        for_containerd=True,
```
This image should no longer be required altogether, right?

NicolasT replied:
Never mind, 'not yet' :)

slaperche-scality committed:
build/constant: move ISO_IMAGE_ROOT to constants

We will need it outside of `image.py`.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
build/salt: generate nginx config for registry

Since the `static-container-registry` tool is written in Python, we can
import it and call the right function instead of calling the tool as a
shell command, which avoid:
- handling stdout redirection (the tool can't write into a file)
- a fork to spawn a subcommand

Note that we have to use `importlib`, because we can't use the `import`
statement with names containing hyphens.

The call to `static-container-registry` is wrapped into a class that
inherits `targets.FileTarget` so that we can use it in the
`targets.FileTree`.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.1/feature/1049-rework-offline-image-serving` with contents from `feature/1049-rework-offline-image-serving`
and `development/2.1`.

Please resolve the conflict on **the integration branch** (`w/2.1/feature/1049-rework-offline-image-serving`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.1/feature/1049-rework-offline-image-serving
 $ git pull  # or "git reset --hard origin/w/2.1/feature/1049-rework-offline-image-serving"
 $ git merge origin/development/2.1
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/feature/1049-rework-offline-image-serving
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/1049-rework-offline-image-serving
```

slaperche-scality committed:
build/salt: fix schema in docstring

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt: deploy nginx config for container registry

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #7cab6d9248b6e2487ed798b23f9e8ccc9d0a1262 on the integration branch
`w/2.1/feature/1049-rework-offline-image-serving` is merging a branch which is neither the current
branch `feature/1049-rework-offline-image-serving` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1049-rework-offline-image-serving` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality committed:
salt: get rid of the registry

We now fetch the images from nginx.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt: inject nginx image

Otherwise you'll have a chicken or the egg problem ^^

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt: fix nginx configuration

Rename the new configuration as `.inc` instead of `.conf` to avoid them
to be automagically included by the `/etc/nginx/nginx.conf` in the
container (it contains `include /etc/nginx/conf.d/*.conf;`) which result
in error (because `set` and `location` are then included out of a
`server` block…).

And make sure that the configuration and the images are mounted in the
container.

Also fix the content of 90-metalk8s-registry-config.inc.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
rename 'package-repositories' to 'repositories'

The repositories now also serve images, not only repositories.

Closes: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on buildchain/buildchain/targets/remote_image.py:
```diff
@@ -77,54 +73,94 @@ def fullname(self) -> str:
-        return actions
+
+
+class RemoteTarImage(base.FileTarget):
```
So, not handling #609 in this PR?

slaperche-scality replied:
I tried to, using `skopeo copy … docker-archive:path/to/image` + using `ctr -n k8s.io image import path/to/image` but bootstrap wasn't working. I'll look at the logs tomorrow, to see if it's easy to fix or not.

slaperche-scality replied:
Done now :)

NicolasT commented on salt/metalk8s/repo/installed.sls:
```diff
@@ -19,27 +27,64 @@ Generate package repositories nginx configuration:
-      - metalk8s: Install package repositories manifest
+      - metalk8s: Install repositories manifest
+
+Deploy container registry nginx configuration:
```
Shouldn't this happen before we deploy the service manifest? I'm aware the `onchanges` acts as a `require`, so things will be executed by Salt 'in the right order', but I think it also makes it easier to read/follow an SLS when tasks are 'in the expected order of execution' within the source.

Same applies to the step below.

slaperche-scality replied:
Done, I moved `Deploy container registry nginx configuration` and `Generate container registry configuration` before `Install repositories manifest`

NicolasT reviewed (dismissed):
One minor request, just moving two blocks up a couple of lines. Other than that LGTM. CI appears to be failing though? Didn't look at logs.

slaperche-scality committed:
lint: fix pylint dict for CI

CI probably use a different version of the `enchant` C library (and thus
comes with different default dictionary).

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt: use `ctr image import` instead of `ctr cri load`

That way, we can generate the archive using skopeo instead of docker (we
can now remove some code from `docker_command`).

Note that since the new command doesn't output anything on stdout, we
can no longer extract the image name from here: we use the filename
instead to populate `changes`.

Refs: #1049
Closes: #609
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

jbertran commented on salt/metalk8s/repo/installed.sls:
```diff
@@ -61,12 +66,13 @@ Install package repositories manifest:
     - context:
         container_port: {{ repo.port }}
-        image: {{ package_repositories_image }}
+        image: docker.io/library/nginx:1.15.8
```
Could take the `alpine` image for this, since it's uncompressed - just append `-alpine`. Isofunctional as far as a `doit vagrant_up` is concerned.

slaperche-scality replied:
As discussed on the channel, we should try to be compliant and use CentOS, so no Alpine.

jbertran commented on salt/metalk8s/repo/files/nginx.conf.j2:
```diff
@@ -6,4 +6,7 @@ server {
     }
+
+    include conf.d/{{ nginx_registry_config }};
+    include conf.d/{{ nginx_registry }};
```
minor: `s/nginx_registry/nginx_registry_images` for clarity?

slaperche-scality replied:
I'm now including `*.inc*.

jbertran commented on salt/metalk8s/repo/installed.sls:
```diff
@@ -6,17 +6,18 @@
+
+{%- set nginx_confd  = '/var/lib/metalk8s/package-repositories/conf.d/' %}
+{%- set nginx_default_conf = nginx_confd ~ 'default.conf' %}
+{%- set nginx_registry        = '99-' ~ saltenv ~ '-registry.inc' %}
```
Ditto for the actual image conf filename, for clarity.

slaperche-scality committed:
build: get rid of `FileTarget` abstraction

In the long run, this has shown itself to be a wrong one that forced me
to jump through hoops (`RemoveTarImage`) so let's get rid of it and use
another approach.

The role of this class was to enforce targets with a single output file
(in order to use them in the `FileTree` class): let's add
`_get_destination` free function in `file_tree.py` that will dynamically
check if the target respect our contract (this will fail at the
"compile" time of the build, not during the build so it's fine).

As a side effect, we can now get rid of the ugly `RemoveTarImage`.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

jbertran commented on salt/metalk8s/repo/installed.sls:
```diff
@@ -2,20 +2,20 @@
-{%- set package_repositories_image = build_image_name('nginx', '1.15.8') %}
+{%- set repositories_name = 'repositories' %}
+{%- set repositories_version = '1.0.0' %}
+{%- set repositories_image = build_image_name('nginx', '1.15.8') %}
```
If we change the image to alpine, should change here too?

jbertran reviewed (commented):
LGTM. I'll do a pass on all pulled images to see if there are any lighter alternatives since we're at it.

slaperche-scality committed:
salt: fix macro for prometheus

States for prometheus were recently added and thus were still
referencing the old location of the macro.

Closes: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
Squashed 'buildchain/static-container-registry/' content from commit c7574927

git-subtree-dir: buildchain/static-container-registry
git-subtree-split: c757492721147f4eab620b8a389eb25c01035f37

slaperche-scality committed:
Merge commit 'e4d10d219990ce7267e09e26e2e555cc3e498539' as 'buildchain/static-container-registry'

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/1049-rework-offline-image-serving` with contents from `feature/1049-rework-offline-image-serving`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/1049-rework-offline-image-serving origin/development/2.1
 $ git merge origin/feature/1049-rework-offline-image-serving
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/1049-rework-offline-image-serving
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/feature/1049-rework-offline-image-serving` with contents from `w/2.1/feature/1049-rework-offline-image-serving`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/feature/1049-rework-offline-image-serving origin/development/2.2
 $ git merge origin/w/2.1/feature/1049-rework-offline-image-serving
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/feature/1049-rework-offline-image-serving
```

slaperche-scality committed:
ci: add new dependencies in the CI

skopeo is used to save the images.
hardlinks is used to deduplicate the layers.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

slaperche-scality committed:
build: honor `DOCKER_HOST` with skopeo

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# History mismatch



Merge commit #6c632cf7851a5bb38433ef7a4627b39b45df0685 on the integration branch
`w/2.1/feature/1049-rework-offline-image-serving` is merging a branch which is neither the current
branch `feature/1049-rework-offline-image-serving` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1049-rework-offline-image-serving` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**

bert-e commented:
# History mismatch



Merge commit #6c632cf7851a5bb38433ef7a4627b39b45df0685 on the integration branch
`w/2.1/feature/1049-rework-offline-image-serving` is merging a branch which is neither the current
branch `feature/1049-rework-offline-image-serving` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1049-rework-offline-image-serving` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #89f1ce60f777e79b591ee1eff378daa9588e9b0d on the integration branch
`w/2.1/feature/1049-rework-offline-image-serving` is merging a branch which is neither the current
branch `feature/1049-rework-offline-image-serving` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1049-rework-offline-image-serving` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**

bert-e commented:
# History mismatch



Merge commit #89f1ce60f777e79b591ee1eff378daa9588e9b0d on the integration branch
`w/2.1/feature/1049-rework-offline-image-serving` is merging a branch which is neither the current
branch `feature/1049-rework-offline-image-serving` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1049-rework-offline-image-serving` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

slaperche-scality commented:
@bert-e force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/1049-rework-offline-image-serving` with contents from `feature/1049-rework-offline-image-serving`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/1049-rework-offline-image-serving origin/development/2.1
 $ git merge origin/feature/1049-rework-offline-image-serving
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/1049-rework-offline-image-serving
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/feature/1049-rework-offline-image-serving` with contents from `w/2.1/feature/1049-rework-offline-image-serving`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/feature/1049-rework-offline-image-serving origin/development/2.2
 $ git merge origin/w/2.1/feature/1049-rework-offline-image-serving
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/feature/1049-rework-offline-image-serving
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

gdemonet commented on buildchain/buildchain/targets/image.py:
```diff
@@ -31,41 +29,43 @@ def __init__(
-        return filename
+    def dirname(self) -> Path:
+        """Directory where to store the image on disk."""
+        return self.dest_dir/self.tag.replace(':', '/')
```
Is the `:` forbidden in an image name? (would make sense)

In any case, unless we want to magically create subdirectories if an image name contains colons, using `self.dest_dir/self.name/self.version` could look more explicit IMHO.

slaperche-scality replied:
You're right, better be explicit here.

gdemonet commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -167,41 +163,41 @@ def check_dockerfile_dependencies(self) -> Union[TaskError, bool]:
+                'docker-daemon:{}'.format(self.tag),
+                'dir:{}'.format(str(self.dirname))
+            ])
+            task['targets'] = [self.dirname/'manifest.json']
```
I'm curious: why do you overwrite `task['targets']` here, while you just `append` below?

slaperche-scality replied:
No reason, I'll fix it.

gdemonet commented on buildchain/buildchain/targets/local_image.py:
```diff
@@ -167,41 +163,41 @@ def check_dockerfile_dependencies(self) -> Union[TaskError, bool]:
         return task
 
-    def _build_actions(self) -> List[types.Action]:
+    def _build_actions(self, task: types.TaskDict) -> None:
```
Maybe we could rename this method (it's not just creating the `actions` actually), or just merge it with the `task` property?

slaperche-scality replied:
I'll rework this part (either rename or merge, depends on what's clearer)

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh

gdemonet reviewed (dismissed):
Really minor comments that can be ignored, let's merge this :)

slaperche-scality committed:
build: cleanup the code

Make `dirname` more explicit and rework `_build_actions` so that it only
build actions.

Refs: #1049
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @cmonfort

* @jbertran

* @monpote

* @alxf

* @gdemonet

* @lucieleonard

* @zempashi

* @carlito-scality

* @alexandre-allard-scality

* @nootal

* @nicolast

* @thomasdanan

* @tcarmet

* @slaperche-scality

* @teddyandrieux

* @sayf-eddine-scality

* @ramanduh










*The following options are set:* **approve**

gdemonet reviewed (approved):
LGTM

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
